src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionService.CacheEntry.cs

```diff
@@ -0,0 +1,151 @@
+                        _complexItemBuilder.ToImmutable());
+                }
+
+                public void AddItem(SyntaxTreeIndex syntaxIndex)
+                {
+                    foreach (var (targetType, symbolInfoIndices) in syntaxIndex.SimpleExtensionMethodInfo)
+                    {
+                        foreach (var index in symbolInfoIndices)
+                        {
+                            _simpleItemBuilder.Add(targetType, syntaxIndex.DeclaredSymbolInfos[index]);
+                        }
+                    }
+
+                    foreach (var index in syntaxIndex.ComplexExtensionMethodInfo)
+                    {
+                        _complexItemBuilder.Add(syntaxIndex.DeclaredSymbolInfos[index]);
+                    }
+                }
+
+                public void Dispose()
+                    => _complexItemBuilder.Free();
+            }
+        }
+
+        /// <summary>
+        /// We don't use PE cache from the service, so just pass in type `object` for PE entries.
+        /// </summary>
+        [ExportWorkspaceServiceFactory(typeof(IImportCompletionCacheService<CacheEntry, object>), ServiceLayer.Editor), Shared]
+        private sealed class CacheServiceFactory : AbstractImportCompletionCacheServiceFactory<CacheEntry, object>
+        {
+            [ImportingConstructor]
+            public CacheServiceFactory()
+            {
+            }
+        }
+
+        private static IImportCompletionCacheService<CacheEntry, object> GetCacheService(Workspace workspace)
+            => workspace.Services.GetRequiredService<IImportCompletionCacheService<CacheEntry, object>>();
+
+        private static async Task<CacheEntry?> GetCacheEntryAsync(
+            Project project,
+            bool loadOnly,
+            IImportCompletionCacheService<CacheEntry, object> cacheService,
+            CancellationToken cancellationToken)
+        {
+            // While we are caching data from SyntaxTreeInfo, all the things we cared about here are actually based on sources symbols.
+            // So using source symbol checksum would suffice.
+            var checksum = await SymbolTreeInfo.GetSourceSymbolsChecksumAsync(project, cancellationToken).ConfigureAwait(false);
+
+            // Cache miss, create all requested items.
+            if (!cacheService.ProjectItemsCache.TryGetValue(project.Id, out var cacheEntry) ||
+                cacheEntry.Checksum != checksum ||
+                cacheEntry.Language != project.Language)
+            {
+                var syntaxFacts = project.LanguageServices.GetRequiredService<ISyntaxFactsService>();
+                using var builder = new CacheEntry.Builder(checksum, project.Language, syntaxFacts.StringComparer);
+
+                foreach (var document in project.Documents)
+                {
+                    // Don't look for extension methods in generated code.
+                    if (document.State.Attributes.IsGenerated)
```

Comment:
we should actually use the generated-code service.  i believe this actual attribute is a *subset* of cases we care about.  it shouldn't be interrogated directly.

Reply:
>  i believe this actual attribute is a subset of cases we care about.

i think this is OK, as this is only intended as a perf optimization. Looking at the implementation of the service, it might not be much more expensive to just check/create syntax index, which would defeat its original purpose
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/GeneratedCodeRecognition/AbstractGeneratedCodeRecognitionService.cs,fdbf7fe3da0eadb2,references